File: built-in plugins dsp/mb-strings-legato.cpp
/**
 * MB Legato Strings
 * Category : instrument
 * Type     : strings
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Smooth connected legato strings
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_STRINGS_LEGATO_H
#define MB_STRINGS_LEGATO_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbStringsLegato : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-strings-legato";
    static constexpr const char* PLUGIN_NAME    = "MB Legato Strings";
    static constexpr const char* PLUGIN_TYPE    = "strings";
    static constexpr const char* PLUGIN_CATEGORY = "instrument";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float glide = 0.3f;  // range [0, 1]
    float volume = 0.8f;  // range [0, 1]
    };

    MbStringsLegato() = default;
    ~MbStringsLegato() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.glide = std::clamp(params.glide, 0f, 1f);
        params.volume = std::clamp(params.volume, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Legato Strings
        return input;
    }
};

#endif // MB_STRINGS_LEGATO_H

File: built-in plugins dsp/mb-strings-legato.ts
import type { PluginDefinition } from "../server/services/pluginHostService";

const MbStringsLegatoPlugin: PluginDefinition = {
  id: "mb-strings-legato",
  slug: "mb-strings-legato",
  name: "MB Legato Strings",
  category: "instrument",
  type: "strings",
  version: "1.0.0",
  description: "Smooth connected legato strings",
  author: "Max Booster",
  grade: "A",
  oscillators: [
    { type: "sawtooth", detune: -7, gain: 0.25 },
    { type: "sawtooth", detune: 7, gain: 0.25 },
    { type: "triangle", detune: 0, gain: 0.5 },
  ],
  envelope: { attack: 0.5, decay: 0.3, sustain: 0.9, release: 1.5 },
  parameters: [
    {
      id: "glide",
      name: "Glide",
      type: "float",
      defaultValue: 0.3,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
    {
      id: "volume",
      name: "Volume",
      type: "float",
      defaultValue: 0.8,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
  ],
  defaultPreset: { glide: 0.3, volume: 0.8 },
};

export default MbStringsLegatoPlugin;
